near-sdk/src/store/unordered_set/mod.rs

```diff
@@ -0,0 +1,798 @@
+    index: LookupMap<T, FreeListIndex, H>,
+}
+
+//? Manual implementations needed only because borsh derive is leaking field types
+// https://github.com/near/borsh-rs/issues/41
+impl<T, H> BorshSerialize for UnorderedSet<T, H>
+where
+    T: BorshSerialize + Ord,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    fn serialize<W: borsh::maybestd::io::Write>(
+        &self,
+        writer: &mut W,
+    ) -> Result<(), borsh::maybestd::io::Error> {
+        BorshSerialize::serialize(&self.elements, writer)?;
+        BorshSerialize::serialize(&self.index, writer)?;
+        Ok(())
+    }
+}
+
+impl<T, H> BorshDeserialize for UnorderedSet<T, H>
+where
+    T: BorshSerialize + Ord,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    fn deserialize(buf: &mut &[u8]) -> Result<Self, borsh::maybestd::io::Error> {
+        Ok(Self {
+            elements: BorshDeserialize::deserialize(buf)?,
+            index: BorshDeserialize::deserialize(buf)?,
+        })
+    }
+}
+
+impl<T, H> Drop for UnorderedSet<T, H>
+where
+    T: BorshSerialize + Ord,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    fn drop(&mut self) {
+        self.flush()
+    }
+}
+
+impl<T, H> fmt::Debug for UnorderedSet<T, H>
+where
+    T: BorshSerialize + Ord + BorshDeserialize + fmt::Debug,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("UnorderedSet")
+            .field("elements", &self.elements)
+            .field("index", &self.index)
+            .finish()
+    }
+}
+
+impl<T> UnorderedSet<T, Sha256>
+where
+    T: BorshSerialize + Ord,
+{
+    #[inline]
+    pub fn new<S>(prefix: S) -> Self
+    where
+        S: IntoStorageKey,
+    {
+        Self::with_hasher(prefix)
+    }
+}
+
+impl<T, H> UnorderedSet<T, H>
+where
+    T: BorshSerialize + Ord,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    /// Initialize a [`UnorderedSet`] with a custom hash function.
+    ///
+    /// # Example
+    /// ```
+    /// use near_sdk::crypto_hash::Keccak256;
+    /// use near_sdk::store::UnorderedMap;
+    ///
+    /// let map = UnorderedMap::<String, String, Keccak256>::with_hasher(b"m");
+    /// ```
+    pub fn with_hasher<S>(prefix: S) -> Self
+    where
+        S: IntoStorageKey,
+    {
+        let mut vec_key = prefix.into_storage_key();
+        let map_key = [vec_key.as_slice(), b"m"].concat();
+        vec_key.push(b'v');
+        Self { elements: FreeList::new(vec_key), index: LookupMap::with_hasher(map_key) }
+    }
+
+    /// Returns the number of elements in the set.
+    pub fn len(&self) -> u32 {
+        self.elements.len()
+    }
+
+    /// Returns true if the set contains no elements.
+    pub fn is_empty(&self) -> bool {
+        self.elements.is_empty()
+    }
+
+    /// Clears the set, removing all values.
+    pub fn clear(&mut self)
+    where
+        T: BorshDeserialize + Clone,
+    {
+        for e in self.elements.drain() {
+            self.index.remove(&e);
```

Comment:
```suggestion
            self.index.set(e, None);
```

remove reads existing value from storage, which I don't think can be compiled away because it uses a host function to do so